src/ring_buffer.zig

```diff
@@ -0,0 +1,82 @@
+const std = @import("std");
+
+/// A First In/First Out ring buffer holding at most `size` elements.
+pub fn RingBuffer(comptime T: type, comptime size: usize) type {
+    return struct {
+        const Self = @This();
+
+        buffer: [size]T = undefined,
+        /// The index of the slot with the first item, if any.
+        index: usize = 0,
+        /// The number of items in the buffer.
+        count: usize = 0,
+
+        /// Add an element to the RingBuffer. Returns an error if the buffer
+        /// is already full and the element could not be added.
+        pub fn push(self: *Self, item: T) error{NoSpaceLeft}!void {
+            if (self.is_full()) return error.NoSpaceLeft;
+            self.buffer[(self.index + self.count) % self.buffer.len] = item;
+            self.count += 1;
+        }
+
+        /// Return but do not remove the next item, if any.
+        pub fn peek(self: *Self) ?T {
+            if (self.is_empty()) return null;
+            return self.buffer[self.index];
+        }
+
+        /// Remove and return the next item, if any.
+        pub fn pop(self: *Self) ?T {
+            if (self.is_empty()) return null;
+            const ret = self.buffer[self.index];
+            self.index = (self.index + 1) % self.buffer.len;
+            self.count -= 1;
+            return ret;
+        }
+
+        pub fn is_full(self: Self) bool {
+            return self.count == self.buffer.len;
+        }
+
+        pub fn is_empty(self: Self) bool {
+            return self.count == 0;
+        }
```

Comment:
> Does the context need to be `*Self`?

Nope, we don't need `self` to be mutable here and zig will optimize the pass-by-value to a pass-by-pointer in this case.

> We could also call these just `full()` or `empty()` since they couldn't be getters or setters otherwise. I only did `is_repair()` in VR because I couldn't come up with another name there.

Fine by me.

Reply:
Somehow I managed to edit and overwrite your comment instead of replying, should be fixed now.

Reply:
> Nope, we don't need self to be mutable here and zig will optimize the pass-by-value to a pass-by-pointer in this case.

Ah, great to know that. Makes perfect sense, just like when using a `for` loop.